Ficr prog sample (#8952)
nrfconnect/sdk-nrf: samples/wifi/radio_test/ficr.rst

.. _wifi_ficr_prog:

Wi-Fi: FICR programming
#######################

.. contents::
   :local:
   :depth: 2

> The Wi-Fi Radio Factory Information Configuration Registers (FICR) programming
> sample demonstrates how to program the user region of FICR parameters on the
> development kit using a set of predefined commands. Please note that all the
> FICR registers are stored in a One Time Programmable (OTP) memory and
> consequently, write commands are DESTRUCTIVE, and once written, the contents
> of OTP registers cannot be re-programmed.

Reviewer: Moved the second sentence to the Overview section as a Note.
```suggestion
The Wi-Fi Radio Factory Information Configuration Registers (FICR) programming sample demonstrates how to program the user region of FICR parameters on the development kit. It uses a set of predefined commands.
```

Please read the FICR section of the product specifications carefully before

> using using these commands.

Author: Yes, it is very important that the person using these functions refers to the register definitions in the product specs before attempting OTP programming

> using using these commands.

Reviewer: Ok, then I believe we can add this line in the Note (already there in the Overview section) as well.

Requirements
************

The sample supports the following development kit:

.. table-from-sample-yaml::

Overview
********

To run the commands, connect to the development kit through the serial port and send shell commands.
Zephyr's :ref:`zephyr:shell_api` module is used to handle the commands.

> You can start running ``wifi_radio_ficr_prog`` subcommands to read or write OTP registers

Author: OK done.

See :ref:`wifi_radio_ficr_prog_subcmds` for a list of available subcommands.

User interface
**************

``wifi_radio_ficr_prog`` is the Wi-Fi radio ficr_prog command and it supports the following subcommands:

.. _wifi_radio_ficr_prog_subcmds:

Wi-Fi radio FICR subcommands
============================

.. list-table:: Wi-Fi radio ficr subcommands
   :widths: 15 15 10 30 70
   :header-rows: 1

   * - Subcommand
     - Register
     - Offset
     - Argument(s)
     - Description
   * - otp_get_status
     - N/A
     - N/A
     - N/A

>      - reads out the OTP status of user region and status of each field (programmed or not)

Author: OK done.

   * - otp_read_params
     - N/A
     - N/A
     - N/A

>      - reads out all the OTP parameters (excluding QSPI_KEY which cannot be read)

Author: OK done.

   * - otp_write_params
     - REGION_PROTECT
     - 0x100
     - arg
     - | arg = 0x50FA50FA : Enable R/W permission
       | arg = 0x00000000 : Lock Region from Writing
       | arg = Others : Invalid
       | All four REGION_PROTECT registers are written with this 32-bit argument
   * - otp_write_params
     - QSPI_KEY
     - 0x110
     - arg1 arg2 arg3 arg4

>      - All 4 key arguments are 32-bit values forming the required 128-bit (Q)SPI key for encryption.

Author: OK done.

   * - otp_write_params
     - MAC_ADDRESS0
     - 0x120
     - arg1 arg2
     - If MAC addr is AA:BB:CC:DD:EE:FF, then arg1=0xDDCCBBAA and arg2=0xFFEE
   * - otp_write_params
     - MAC_ADDRESS1
     - 0x128
     - arg1 arg2

>      - If MAC addr is AA:BB:CC:DD:EE:FF, then arg1=0xDDCCBBAA and arg2=0xFFEE

Author: OK done.

   * - otp_write_params
     - CALIB_XO
     - 0x130
     - arg

>      - arg is 7-bit unsigned XO value. Bits [31:7] unused. Adjusts capacitor bank, 0 : Lowest capacitance (Highest frequency), 255 : Highest capacitance (Lowest frequency)

Author: OK done.

   * - otp_write_params
     - CALIB_PDADJMCS7
     - 0x134
     - arg

>      - arg is 32-bit power detector adjustment for MCS7 for channel 7, 36, 100, 165 (channel 7 in bits[7:0] - signed)

Author: OK done.

>      - arg is 32-bit power detector adjustment for MCS7 for channel 7, 36, 100, 165 (channel 7 in bits[7:0] - signed)

Reviewer: I guess it's missed.

   * - otp_write_params
     - CALIB_PDADJMCS0
     - 0x138
     - arg

>      - arg is 32-bit power detector adjustment for MCS0 for channel 7, 36, 100, 165 (channel 7 in bits[7:0] - signed)

Author: OK done.

>      - arg is 32-bit power detector adjustment for MCS0 for channel 7, 36, 100, 165 (channel 7 in bits[7:0] - signed)

Reviewer: I guess it's missed.

   * - otp_write_params
     - CALIB_MAXPOW2G4
     - 0x13c
     - arg

>      - arg is 32-bit value specifying Max output power for DSSS, MCS0, MCS7 (DSSS in bits[7:0] - unsigned). Measured in channel 7. Bits [31:24] unused.

Author: OK done.

   * - otp_write_params
     - CALIB_MAXPOW5G0MCS7
     - 0x140
     - arg

>      - arg is 32-bit value specifying Max output power for MCS7 in channel 36, 100, 165 (channel 36 in bits[7:0] - unsigned). Bits [31:24] unused.

Author: OK done.

   * - otp_write_params
     - CALIB_MAXPOW5G0MCS0
     - 0x144
     - arg

>      - arg is 32-bit value specifying Max output power for MCS0 in channel 36, 100, 165 (channel 36 in bits[7:0] - unsigned). Bits [31:24] unused.

Author: OK done.

   * - otp_write_params
     - CALIB_RXGAINOFFSET
     - 0x148
     - arg
     - arg is 32-bit value specifying RX gain offset for channel 7, 36, 100, 165 (channel 7 in bits[7:0] - signed)
   * - otp_write_params
     - CALIB_TXPOWBACKOFFT
     - 0x14C
     - arg
     - arg is 32-bit value specifying TX power backoff versus temperature. Bits[7:0] 2.4G at 80C, bits[15:8] 2.4G at -20C, bits[23:16] 5.0G at 80C, bits[31:24] 5.0G at -20C. Each 8-bit field is signed value
   * - otp_write_params
     - CALIB_TXPOWBACKOFFV
     - 0x150
     - arg
     - arg is 32-bit value specifying TX power backoff versus voltage. Backoff in 2.4GHz, 5GHz bands  at "Very low" voltage. Backoff in 2.4GHz, 5GHz bands  at "Low" voltage. Each 8-bit field is signed value
   * - otp_write_params
     - REGION_DEFAULTS
     - 0x154
     - arg
     - | Register default enable control.

>        | arg is 32-bit value where a specific bit set to '0' implies the corresponding OTP field is programmed. The bit & field mapping are

Author: OK done.

       |
       | bit 0  : QSPI_KEY
       | bit 1  : MAC0 Address
       | bit 2  : MAC1 Address
       | bit 3  : CALIB_XO
       | bit 4  : CALIB_PDADJMCS7
       | bit 5  : CALIB_PDADJMCS0
       | bit 6  : CALIB_MAXPOW2G4
       | bit 7  : CALIB_MAXPOW5G0MCS7
       | bit 8  : CALIB_MAXPOW5G0MCS0
       | bit 9  : CALIB_RXGAINOFFSET
       | bit 10 : CALIB_TXPOWBACKOFFT
       | bit 11 : CALIB_TXPOWBACKOFFV
       | bit 12-31 : Reserved

Building and running
********************

> .. |sample path| replace:: :file:`samples/wifi/radio_test`

Reviewer: ```suggestion
.. |sample path| replace:: :file:`samples/wifi/ficr`
```

Author: OK done.

.. include:: /includes/build_and_run.txt

Currently, the following configurations are supported:

* 7002 DK + QSPI
* 7002 EK + SPIM

To build for the nRF7002 DK, use the ``nrf7002dk_nrf5340_cpuapp`` build target.
The following is an example of the CLI command:

.. code-block:: console

   west build -b nrf7002dk_nrf5340_cpuapp

To build for the nRF7002 EK and nRF5340 DK, use the ``nrf5340dk_nrf5340_cpuapp`` build target with the ``SHIELD`` CMake option set to ``nrf7002_ek``.
The following is an example of the CLI command:

.. code-block:: console

   west build -b nrf5340dk_nrf5340_cpuapp -- -DSHIELD=nrf7002_ek

See also :ref:`cmake_options` for instructions on how to provide CMake options.

Testing
=======

|test_sample|

#. |connect_kit|
#. |connect_terminal|

> #. Reference command interface

Author: OK done.

      .. code-block:: console

          wifi_radio_ficr_prog <subcommand> [Offset] [arg1] [arg2] .. [argN]

> #. Example commands:

Author: OK done.

          [00:01:09.034,881] <inf> otp_prog:
          [00:01:09.034,881] <inf> otp_prog: CALIB_PDADJMCS7 = 0xffffffff
          [00:01:09.034,881] <inf> otp_prog:
          [00:01:09.034,912] <inf> otp_prog: CALIB_PDADJMCS0 = 0xffffffff
          [00:01:09.034,912] <inf> otp_prog:
          [00:01:09.034,942] <inf> otp_prog: CALIB_MAXPOW2G4 = 0xffffffff
          [00:01:09.034,942] <inf> otp_prog:
          [00:01:09.034,942] <inf> otp_prog: CALIB_MAXPOW5G0MCS7 = 0xffffffff
          [00:01:09.034,973] <inf> otp_prog:
          [00:01:09.034,973] <inf> otp_prog: CALIB_MAXPOW5G0MCS0 = 0xffffffff
          [00:01:09.034,973] <inf> otp_prog:
          [00:01:09.035,003] <inf> otp_prog: CALIB_RXGAINOFFSET = 0xffffffff
          [00:01:09.035,003] <inf> otp_prog:
          [00:01:09.035,003] <inf> otp_prog: CALIB_TXPOWBACKOFFT = 0xffffffff
          [00:01:09.035,034] <inf> otp_prog:
          [00:01:09.035,034] <inf> otp_prog: CALIB_TXPOWBACKOFFV = 0xffffffff
          [00:01:09.035,034] <inf> otp_prog:
          [00:01:09.035,064] <inf> otp_prog: REGION_DEFAULTS = 0xfffffff1
          [00:01:09.035,064] <inf> otp_prog:

   #. To read the status of OTP Region

>       Execute the following command:

Author: OK done.

        [00:00:11.178,192] <inf> otp_prog:
        [00:00:11.178,222] <inf> otp_prog: QSPI Keys are not programmed in OTP
        [00:00:11.178,222] <inf> otp_prog: MAC0 Address is programmed in OTP
        [00:00:11.178,222] <inf> otp_prog: MAC1 Address is programmed in OTP
        [00:00:11.178,253] <inf> otp_prog: CALIB_XO is programmed in OTP
        [00:00:11.178,253] <inf> otp_prog: CALIB_PDADJMCS7 not programmed in OTP
        [00:00:11.178,253] <inf> otp_prog: CALIB_PDADJMCS0 is not programmed in OTP
        [00:00:11.178,283] <inf> otp_prog: CALIB_MAXPOW2G4 is not programmed in OTP
        [00:00:11.178,283] <inf> otp_prog: CALIB_MAXPOW5G0MCS7 is not programmed in OTP
        [00:00:11.178,283] <inf> otp_prog: CALIB_MAXPOW5G0MCS0 is not programmed in OTP
        [00:00:11.178,314] <inf> otp_prog: CALIB_RXGAINOFFSET is not programmed in OTP
        [00:00:11.178,314] <inf> otp_prog: CALIB_TXPOWBACKOFFT is not programmed in OTP
        [00:00:11.178,344] <inf> otp_prog: CALIB_TXPOWBACKOFFV is not programmed in OTP

   #. To write different locations of the OTP, follow the below instructions in conjunction with the tables describing

>       Execute the following sequence of commands to program MAC0 address to F0:CE:36:00:00:4A:

Author: OK done.

       | bit 8  : CALIB_MAXPOW5G0MCS0
       | bit 9  : CALIB_RXGAINOFFSET
       | bit 10 : CALIB_TXPOWBACKOFFT
       | bit 11 : CALIB_TXPOWBACKOFFV
       | bit 12-31 : Reserved

Building and running
********************

.. |sample path| replace:: :file:`samples/wifi/ficr`

.. include:: /includes/build_and_run.txt

Currently, the following configurations are supported:

* 7002 DK + QSPI
* 7002 EK + SPIM

To build for the nRF7002 DK, use the ``nrf7002dk_nrf5340_cpuapp`` build target.
The following is an example of the CLI command:

.. code-block:: console

   west build -b nrf7002dk_nrf5340_cpuapp

To build for the nRF7002 EK and nRF5340 DK, use the ``nrf5340dk_nrf5340_cpuapp`` build target with the ``SHIELD`` CMake option set to ``nrf7002_ek``.
The following is an example of the CLI command:

.. code-block:: console

   west build -b nrf5340dk_nrf5340_cpuapp -- -DSHIELD=nrf7002_ek

See also :ref:`cmake_options` for instructions on how to provide CMake options.

Testing
=======

|test_sample|

#. |connect_kit|
#. |connect_terminal|
#. Use the following reference command interface to read or write the OTP params:

      .. code-block:: console

>           wifi_radio_ficr_prog <subcommand> [Offset] [arg1] [arg2] .. [argN]

Reviewer: Need to check the formatting here.
